worker-workflow/src/main/java/com/github/cafdataprocessing/WorkflowWorker.java

```diff
@@ -0,0 +1,182 @@
+    public void checkHealth(final HealthMonitor healthMonitor)
+    {
+        try {
+            workflowAdminApi.healthCheck();
+        } catch (ApiException e) {
+            LOG.error("Problem encountered when contacting Processing API to check health: ", e);
+            healthMonitor.reportUnhealthy("Processing API communication is unhealthy: " + e.getMessage());
+        }
+    }
+
+    /**
+     * Processes a single document. Retrieving the workflow it refers to, transforming that workflow to a runnable script,
+     * evaluating the document against the workflow to determine where it should be sent to and storing the workflow
+     * on the document so the next worker may re-evaluate the document once it has finished its action.
+     *
+     * @param document the document to be processed.
+     * @throws InterruptedException if any thread has interrupted the current thread
+     * @throws DocumentWorkerTransientException if the document could not be processed
+     */
+    @Override
+    public void processDocument(final Document document) throws InterruptedException, DocumentWorkerTransientException
+    {
+        // Get the worker task properties passed in custom data
+        final ExtractedProperties extractedProperties = CustomDataExtractor.extractPropertiesFromCustomData(document);
+        if(!extractedProperties.isValid()) {
+            LOG.warn("Custom data on document is not valid for this worker. Processing of this document will not " +
+                    "proceed for this worker.");
+            return;
+        }
+        final String projectId = extractedProperties.getProjectId();
+        final long workflowId = extractedProperties.getWorkflowId();
+
+        // Get the specified workflow and transform it to a JavaScript representation
+        TransformWorkflowResult transformWorkflowResult = workflowCache.getTransformWorkflowResult(workflowId, projectId);
+        if(transformWorkflowResult==null) {
+            ReentrantLock transformWorkflowLock = workflowCache.getTransformWorkflowLock(workflowId, projectId);
+            transformWorkflowLock.lock();
+            try {
+                transformWorkflowResult = workflowCache.getTransformWorkflowResult(workflowId, projectId);
+                if (transformWorkflowResult == null) {
+                    transformWorkflowResult = transformWorkflow(document, extractedProperties);
+                    if (transformWorkflowResult == null) {
+                        LOG.warn("Failure during workflow transformation. Processing of this document will not proceed " +
+                                "for this worker.");
+                        return;
+                    }
+                    // Cache the transform result
+                    workflowCache.addTransformWorkflowResult(workflowId, projectId, transformWorkflowResult);
+                }
+            }
+            finally {
+                transformWorkflowLock.unlock();
+            }
+        }
```

Comment:
I had started with LoadingCache but couldn't think of a way to add  specific failures during transformation to a document if the loader was in charge of performing the transformation (it would only have the key information available to it). Will take another look at it.

Reply:
The answer to that is with an exception.  If you can't load the cache element then you can throw an exception which you can catch back in the code where you have the document context.